crates/subgraph/src/apy.rs

```diff
@@ -0,0 +1,777 @@
+    trades: &[Trade],
+    vols: &[VaultVolume],
+    start_timestamp: Option<u64>,
+    end_timestamp: Option<u64>,
+) -> Result<Vec<TokenVaultAPY>, OrderbookSubgraphClientError> {
+    let mut token_vaults_apy: Vec<TokenVaultAPY> = vec![];
+    for vol in vols {
+        // this token vault trades in desc order by timestamp
+        let vault_trades = trades
+            .iter()
+            .filter(|v| {
+                (v.input_vault_balance_change.vault.vault_id.0 == vol.id
+                    && v.input_vault_balance_change.vault.token == vol.token)
+                    || (v.output_vault_balance_change.vault.vault_id.0 == vol.id
+                        && v.output_vault_balance_change.vault.token == vol.token)
+            })
+            .collect::<Vec<&Trade>>();
+
+        // this token vault first trade, indictaes the start time
+        // to find the end of the first day to find the starting capital
+        let first_trade = vault_trades[vault_trades.len() - 1];
+        let first_day_last_trade = vault_trades
+            .iter()
+            .filter(|v| {
+                u64::from_str(&v.timestamp.0).unwrap()
+                    <= u64::from_str(&first_trade.timestamp.0).unwrap() + DAY
+            })
+            .collect::<Vec<&&Trade>>()[0];
+
+        // vaults starting capital at end of first day of its first ever trade
+        let starting_capital = if first_day_last_trade
+            .input_vault_balance_change
+            .vault
+            .vault_id
+            .0
+            == vol.id
+            && first_day_last_trade.input_vault_balance_change.vault.token == vol.token
+        {
+            U256::from_str(
+                &first_day_last_trade
+                    .input_vault_balance_change
+                    .new_vault_balance
+                    .0,
+            )?
+        } else {
+            U256::from_str(
+                &first_day_last_trade
+                    .output_vault_balance_change
+                    .new_vault_balance
+                    .0,
+            )?
+        };
+
+        // the time range for this token vault
+        let mut start = u64::from_str(&first_trade.timestamp.0)?;
+        start_timestamp.inspect(|t| {
+            if start > *t {
+                start = *t;
+            }
+        });
+        let end = end_timestamp.unwrap_or(chrono::Utc::now().timestamp() as u64);
+
+        // this token vault apy
+        let apy = if starting_capital.is_zero() {
+            0_f64
+        } else {
+            // by 4 point decimals
+            let change_ratio = i64::try_from(
+                vol.net_vol
+                    .saturating_mul(I256::from_raw(U256::from(10_000)))
+                    .checked_div(I256::from_raw(starting_capital))
+                    .unwrap_or(I256::ZERO),
+            )? as f64;
+            let time_to_year_ratio = ((end - start) as f64) / YEAR as f64;
+            (change_ratio * time_to_year_ratio) / 100f64
```

Comment:
the equation that i gave in chat was wrong, this should be `change_ratio / time_to_year_ratio`

e.g. if you had +$100 net volume on $1000 starting capital after 6 months then your APY should be 20%

(100 / 1000) / (6 / 12) => 0.1 / 0.5 = 0.2

personally i'd leave that as 0.2 rather than trying to convert to a percentage as you're working with floats here

Reply:
will apply the change